### PR TITLE
TRPL: Drop

### PR DESCRIPTION
**Pull request by steveklabnik**
r? @alexcrichton

### CODE REVIEW

**Comment by alexcrichton** [comment]
Nice! I was almost quite surprised at how short this was, so I'm curious, but what do you think about discussing topics like:

* Common usage patterns of destructors, including: writing FFI bindings which clean up owned resources, making a destructor run in linear time when it would otherwise be recursive (e.g. a BST), or some sort of "bomb" which does something (e.g. a profiling timer that registers info when it goes out of scope.
* Hazards of destructors not possibly running in all circumstances (ala `mem::forget`) and possible intentional use cases of canceling a destructor.
* Interaction between `Drop` and `Copy`
* Interaction of RAII and `Drop`

etc.

**Comment by steveklabnik** [comment]
I wasn't sure how far to go, exactly, and didn't want this to be a TODO :)

I'm open to some of this for now, and more later. Basically, I'm mostly focused on getting tutorials done for friday's release, but some of this will be quick to add and have a lot of value. I'll update it later tonight and ping you again :)

File: src/doc/trpl/drop.md [hunk 1]
```diff
@@ -1,3 +1,67 @@
...
+```text
+BOOM times 100!!!
+BOOM times 1!!!
+```
+
+The TNT goes off before the firecracker does, because it was declared
+afterwards. Last in, first out.
+
+So what is `Drop` good for? Generally, `Drop` is used to clean up any resources
+associated with a `struct` other than memory. For example, the [`Arc<T>`
```
Comment (Gankro) [inline src/doc/trpl/drop.md]: This seems a tad awkward. It is definitely also used for cleaning up memory (std::boxed::*, std::rc::* std::collections::*, ...).

Comment (steveklabnik) [inline src/doc/trpl/drop.md]: yeah, i'll remove that

**Comment by steveklabnik** [comment]
@alexcrichton so, in the interest of getting the basics in before the deadline, I'm thinking about leaving that to future work, and just merging what I have now. What do you think?

**Comment by pnkfelix** [comment]
For future reference for the future work: I do so enjoy examples of using `Drop` to instrument dynamic behavior, like the test I wrote up here ([expurgiated playpen version here](http://is.gd/c3pRYO) too).

  https://raw.githubusercontent.com/rust-lang/rust/master/src/test/run-pass/issue-23338-ensure-param-drop-order.rs

**Comment by pnkfelix** [comment]
On a more serious note: I think Knuth uses an "under construction" icon for sections of his text that he knows will be undergoing expansion or revision in the next edition of his text.  Maybe you should consider doing something similar, so that the reader knows that there is more to be learned on the topic, and that they are encouraged to ask around in the community about it?

**Comment by steveklabnik** [comment]
@pnkfelix i like that idea, for sure.

**Pull request merged**